exporter/signalfx/signalfx.go

```diff
@@ -0,0 +1,326 @@
+		metric := signalFxMetric{
+			metricName:  sanitize(vd.View.Name),
+			metricType:  "gauge",
+			metricValue: float64(data.Value),
+			timestamp:   time.Now(),
+			dimensions:  buildDimensions(row.Tags),
+		}
+		return metric
+	default:
+		e.opts.OnError(errors.New(fmt.Sprintf("aggregation %T is not yet supported", data)))
+		return signalFxMetric{}
+	}
+}
+
+// extractViewData extracts stats data
+func extractViewData(vd *view.Data, e *Exporter) {
+	for _, row := range vd.Rows {
+		metric := e.c.formatMetric(vd.View, row, vd, e)
+		go sendRequest(e, metric)
+	}
+}
+
+func buildDimensions(t []tag.Tag) map[string]string {
+	values := make(map[string]string)
+	for _, t := range t {
+		values[t.Key.Name()] = t.Value
+	}
+	return values
+}
+
+type collector struct {
+	opts Options
+	mu   sync.Mutex // mu guards all the fields.
+
+	// viewData are accumulated and atomically
+	// appended to on every Export invocation, from
+	// stats. These views are cleared out when
+	// Collect is invoked and the cycle is repeated.
+	viewData map[string]*view.Data
+
+	registeredViewsMu sync.Mutex
+
+	registeredViews map[string]string
+}
+
+func (c *collector) addViewData(vd *view.Data) {
+	c.registerViews(vd.View)
+	sig := viewSignature(vd.View.Name, vd.View)
+
+	c.mu.Lock()
+	c.viewData[sig] = vd
+	c.mu.Unlock()
+}
+
+type signalFxMetric struct {
+	metricType  string
+	metricName  string
+	metricValue float64
+	timestamp   time.Time
+	dimensions  map[string]string
+	buckets     []int64
+}
+
+func newCollector(opts Options) *collector {
+	return &collector{
+		opts:            opts,
+		registeredViews: make(map[string]string),
+		viewData:        make(map[string]*view.Data),
+	}
+}
+
+func viewName(namespace string, v *view.View) string {
+	var name string
+	if namespace != "" {
+		name = namespace + "_"
+	}
+	return name + sanitize(v.Name)
+}
+
+func viewSignature(namespace string, v *view.View) string {
+	var buf bytes.Buffer
+	buf.WriteString(viewName(namespace, v))
+	for _, k := range v.TagKeys {
+		buf.WriteString("-" + k.Name())
+	}
+	return buf.String()
+}
+
+// sendRequest sends a package of data containing one metric
+func sendRequest(e *Exporter, data signalFxMetric) {
+	client := sfxclient.NewHTTPSink()
+	client.AuthToken = e.opts.Token
+	ctx := context.Background()
+
+	switch data.metricType {
+	case "gauge":
+		err := client.AddDatapoints(ctx, []*datapoint.Datapoint{
+			sfxclient.GaugeF(data.metricName, data.dimensions, data.metricValue),
+		})
+		if err != nil {
+			e.opts.OnError(errors.New(fmt.Sprintf("Error sending datapoint to SignalFx: %T", err)))
+		}
+	case "cumulative_counter":
+		err := client.AddDatapoints(ctx, []*datapoint.Datapoint{
+			sfxclient.CumulativeF(data.metricName, data.dimensions, data.metricValue),
+		})
+		println("sendind...")
+		if err != nil {
+			e.opts.OnError(errors.New(fmt.Sprintf("Error sending datapoint to SignalFx: %T", err)))
+		}
+	case "cumulative_bucket":
+		bucket := sfxclient.CumulativeBucket{
```

Comment:
@mdubbyap Can you advise on whether this should be used here?

Reply:
CumulativeBucket is a very primitive bucket that emits 3 timeseries, count, sum and sumsquares. I very much doubt this is what we want to do.  If this is trying to turn a histogram into signalfx metrics, we probably want to go the route of creating a new mts with same metric with different dimension identifying each bucket.

Reply:
Ok. On Java, this data type (Distribution aggregation) is returned as not supported. Therefore, I will now use the same strategy here.